### ISSUE TITLE
Lua Script

### ISSUE THREAD

**Issue by ggrigon** [vysheng/tg#497]
Everytime i run my telegram-cli, it's reading the buffer of msgs
But i want to ignore it, because if someone flood it while it's offline, then will be a lot of msgs to interpret.

actually im using a script like that

function on_msg_receive (msg)

      if msg.out then
          return
      end
      if (msg.text=='!help') then
         send_msg (msg.from.print_name, 'my custom msg', ok_cb, false)
     end

But it's flooding me, if for some reason its offline and i send 10x !help, once it`s online it start the flood
There's a way to ignore the buffer, and answer only the new commands?

**Comment by LukeLR** [comment]
Hey! Yes, there is a method to prevent tg-cli from parsing old messages after an outage. I'll explain to you :)

Your lua script should contain a method called on_binlog_replay_end, which it usually does already, when you used the tg-lua.lua example from the repository as template. This method, as it's name says, is called, when the replay of old messages ends. You could therefore define a variable called "active" or something, and set it to zero by default. Do this in the first lines of your script. Then, add an if-clause wrapping all of your contents of on_msg_received(), and set it's conditional to active, so it only executes it's contents when the "active"-variable is true. By default, the script therefore will do nothing, since "active" is false.

Now, in the method on_binlog_replay_end, set the variable "active" to true, so when Telegram is finished syncing missed messages, it will get "active". From this point on, every incoming message would result in on_msg_receive to be called, and since "active" is now true, on_msg_receive executes it's contents. While telegram was syncing the old messages, active was false, and therefore it will do nothing to this messages.

Hope that I could help you ;)

**Comment by ggrigon** [comment]
wowo, thanks a lot for the explanation.. 
even it's well explained, i'm still with doubt

can you show me how to make this script?

i see that part on test.lua

function on_binlog_replay_end ()
  started = 1
  postpone (cron, false, 1.0)
end

Where's the place i should set the active variable?

**Comment by LukeLR** [comment]
No problem! I'll just type some quick example script ;)
```lua
active = false
function on_msg_receive(msg)
  if active == true then
  -- the following code will NOT be executed for messages telegram has missed due to downtime
    if msg.text == 'ping' then 
      send_msg (msg.from.print_name, 'pong', ok_cb, false)
    end
  -- add other things here, but always keep it inside the "if active == true" if-clause.
  -- code outside of the "if active == true" if-clause will be executed when telegram pulls missed messages
  -- code inside of the "if active == true" if-clause will only be executed when new messages arrive
  end -- end of the "if active == true" if-clause
-- Any code here WILL be executed both for messages telegram has missed and new messages
end
function on_binlog_replay_end()
  active = true
end
```

**Comment by ggrigon** [comment]
I used exacly ur example, but it's processing the msgs that i send while its offline.

I've send "ping" twice while it's not running, and once i run it again, it replied me with pong twice

**Comment by SyRenity** [comment]
Another option that I seen used, is just dropping every message that is older then current time:

  -- Before bot was started
  if msg.date < now then
    print("Not valid, old msg")
    return false
  end

**Comment by LukeLR** [comment]
Okay, that's weird... Maybe you should do some debugging? Try to add ```print "Replay ends here!"``` into the ```function on_binlog_replay_end()```. Then quit telegram (if you haven't already), send some messages, and then launch it with your script as parameter as usual. Watch the console output of the cli and look for the "Replay ends here!"-message. Does it appear? And if it does, when? It seems to me like an issue of the telegram cli, since the code looks right. Another interesting debugging option would be to add ```print active``` as the second line right after the ```active = false```-Statement, to check if active is really set to false, and then maybe add ```print active``` as first line in the ```function on_msg_receive()``` (*warning, this might cause a lot of output*) to see the state of ```active``` for each message that is processed.

**Comment by ggrigon** [comment]
Here's the output of add print "Replay ends here!"


Telegram-cli version 1.2.0, Copyright (C) 2013-2015 Vitaly Valtman
Telegram-cli comes with ABSOLUTELY NO WARRANTY; for details type `show_license'.
This is free software, and you are welcome to redistribute it
under certain conditions; type `show_license' for details.
Telegram-cli uses libtgl version 1.2.0
I: config dir=[/home/silver/.telegram-cli]
 *** public key '/etc/telegram-cli/server.pub' loaded successfully
 *** Couldn't open public key file: tg-server.pub
 *** Can not load key tg-server.pub
 *** DC1 'DC' update: 149.154.175.50:443
 *** DC2 'DC' update: 149.154.167.51:443
 *** DC3 'DC' update: 149.154.175.100:443
 *** DC4 'DC' update: 149.154.167.91:443
 *** DC5 'DC' update: 149.154.171.5:443
 *** lua: attempt to call a nil value
Replay ends here!
> [warn] Epoll ADD(1) on fd 0 failed.  Old events were 0; read change was 1 (add); write change was 0 (none): Operation not permitted

After that, on bottom appears the msg that i send while its offline

[18:30]  Guilherme Rigon >>> ping
User Guilherme Rigon online (was online [2015/04/27 18:35:33])
[18:30]  Guilherme Rigon <<< pong
 *** lua: attempt to call a nil value

**Comment by SyRenity** [comment]
@LukeLR It's officially used by telegram-bot project, hence I switched to it too (had your solution before).

**Comment by SyRenity** [comment]
@LukeLR Thinking a bit further, the binlog approach will actually ignore any incoming messages coming during binlog replay including new ones, which is a non-desired effect of course.

**Comment by LukeLR** [comment]
Mhh, @SyRenity, your point is not bad... I think that both methods have their downsides. For myself, I wouldn't trust the time-based detection of older messages, maybe a bit overcautious, but I am just afraid what happens when there is a time difference between sending the message and receiving it, and receiving the message and parsing it, or that the time of the client could have an offset etc. Depending on the accuracy of the local time value, even milliseconds could cause a message not to be parsed, so... I prefer the ```on_binlog_replay_end()```-Method.

@ggrigon: Seems like the variable is not declared, that's why these "attempt to call a nil value"-errors pop up. Does your Script really start with ```active = false```, or is there maybe a typo in that declaration / an unwanted character? To exclude any other influences, you should try to start over with an empty file and only include the methods ```on_msg_receive(msg)``` and ```on_binlog_replay_end()``` as well as ```active = false``` at the beginning of the script. Another option could be to switch the variable name ```active``` for something else, ```replay-end``` could be an example, although this should normally not be the problem. Or a bug in the lua-components of your tg-cli-installation, so maybe checking out the git repository and building the source again could help... Looks very, very weird... If that doesn't work, you could try @SyRenity's method and check if that does the trick...

**Comment by ggrigon** [comment]
I don't think its a a typo, already checked entire script, and yes, i started this one from scratch, based on the one that u share here.

"replay-end" as variable gimme errors, so i changed it to "replayend"

see the output:

script:
cat action.lua
replayend = false

function on_msg_receive(msg)
  if replayend == true then
  -- the following code will NOT be executed for messages telegram has missed due to downtime
    if msg.text == 'ping' then 
      send_msg (msg.from.print_name, 'pong', ok_cb, false)
    end

  -- add other things here, but always keep it inside the "if active == true" if-clause.
  -- code outside of the "if active == true" if-clause will be executed when telegram pulls missed messages
  -- code inside of the "if active == true" if-clause will only be executed when new messages arrive
  end -- end of the "if active == true" if-clause
-- Any code here WILL be executed both for messages telegram has missed and new messages
end



function on_binlog_replay_end() 
replayend = true
print "Replay ends here!"
end

--------------------------------

I: config dir=[/home/silver/.telegram-cli]
 *** lua: attempt to call a nil value
Replay ends here!
 *** lua: attempt to call a nil value

**Comment by LukeLR** [comment]
Okay, @Guilherme, you're probably right. My script does parse missed messages, so I'm not using this method right now, but I'll swap my lua scripts then and try this out, so maybe I can find the bug ;)

**Comment by vysheng** [comment]
ok, I understand. I'll add some key to do this. For now you can do this: 

Insert this code after lua_binlog_end () call in loop.c

	TLS->pts = 0;
	TLS->qts = 0;
	TLS->seq = 0;
	TLS->date = 0;

It's a cheat, but should work. I'll add it properly in new version

**Comment by SyRenity** [comment]
I just added some buffer time (5 seconds) to make sure I don't loose any
pending messages.

**Comment by SyRenity** [comment]
Thanks!

**Comment by vysheng** [comment]
```on_get_difference_end``` function. Maybe it is what you need

**Comment by SyRenity** [comment]
@vysheng Thanks, I will check.

**Comment by vysheng** [comment]
yes. After get_difference () call, that receives messages since last online

**Comment by SyRenity** [comment]
@vysheng however in this case we still have the issue that any messages sent between CLI start and the difference firing, will be filtered out

**Comment by ggrigon** [comment]
@LukeLR 
have u tried anything else?

**Comment by LukeLR** [comment]
No, I haven't found the time yet... Probably tomorrow...

**Comment by michelePap** [comment]
Hello, someone have found a workaround for this issue?
thanks

**Comment by Testato** [comment]
@michelePap 

for me whit vysheng solution it work.
The message is not parsed and also not printed on terminal consolle.

pay attention that you need re-compile after the modification

**Comment by caliph007** [comment]
Its now over a year later.
What is the current recommended way of parsing ony new (unprocessed) chat messages in the lua script?

**Comment by henryaukc** [comment]
Hi, I am a newbie and using telegram to communicate with my raspberry to do stuff. I would like to ask all the method or data member of the object  "msg" passed by the on_receive_message function? I tried hard to find a detail document but can't find it.  I wanna know if the message is a pic or text or even a file and do different things accordingly. Thanks a lot for your help.

**Comment by ggrigon** [comment]
@henryaukc 

this one?

`function on_msg_receive(msg)
`

it was for msg text on my code
